assembly/src/library/path.rs

```diff
@@ -18,18 +18,18 @@ use crate::{
 };
 
 /// Represents errors that can occur when creating, parsing, or manipulating [LibraryPath]s
-#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+#[derive(Debug, thiserror::Error)]
 pub enum PathError {
     #[error("invalid library path: cannot be empty")]
     Empty,
     #[error("invalid library path component: cannot be empty")]
     EmptyComponent,
     #[error("invalid library path component: {0}")]
-    InvalidComponent(#[from] crate::ast::IdentError),
```

Comment:
Why are we removing the `#[from]`? Using `map_err()` everywhere can get pretty verbose, although it is a matter of taste

Reply:
This is a matter of taste, yes, but I think there is a tendency to produce worse errors when using `From`. For example, you may have a `LibraryError` wrapped in an `AssemblyError`, which probably looks like `AssemblyError::Library(#[from] LibraryError)`. The problem with this is that it doesn't add any context and is too generic. What is better, I think, is to have a dedicated error variant for the each error case where a library error must be wrapped into an assembly error. Imaginary examples:

```rust
enum AssemblyError {
   #[error("failed to assemble kernel library")]
    KernelLibraryAssemblyFailed(#[source] LibraryError),
   #[error("failed to load library")]
    FailedToLoadLibrary(#[source] LibraryError),
}
```

So it's possible to have `LibraryError` in `AssemblyError` multiple times and adding `#[from]` in this case produces conflicting implementations.
The lack of `From` is good, because it forces the developer to use `map_err` and think about what error case to map the error to, rather than bubbling it up blindly. This tends to lead to better errors.

Of course, it's not always wrong to use `#[from]`, in simple cases it's fine, especially when you can't add meaningful context. I personally like to avoid it generally because it simply forces me to think about the error context more carefully and that's a good thing.

Ultimately you have worked more in this repo than I did, so feel free to make a call on whether to use it or not :slightly_smiling_face:, but that is the rationale.

Reply:
Yeah no this makes sense - the theory being that e.g. writing `.map_err(FailedToLoadLibrary)` would make you realize that this is not quite the variant you want, and make you create the new `KernelLibraryAssemblyFailed` instead. I'm good to try this moving forward.

Reply:
@PhilippGackstatter The way `thiserror` works already ensures that context is captured via backtraces (and when combined with `anyhow`, allows additional ad-hoc context to be added). The `Report` type specifically already does both (and includes even richer diagnostic information like associated source code spans).

I find that it is rarely the case that manually constructing errors versus using `#[from]` produces any better error information in practice. If people aren't adding context either way, you end up with the same outcome.

In general, at least with the assembler, the use of `Report` is encouraged/recommended to ensure that it is easier to do this, and to avoid a bunch of useless error variants being defined in favor of ad-hoc messages and site-specific context, but there is still a place for the former, when you simply need to know the type of error (and optionally, the backtrace).

Anyway, wanted to add my two cents to this.